src/scimlfunctions.jl

```diff
@@ -1537,23 +1537,27 @@ OptimizationFunction{iip}(f,adtype::AbstractADType=NoAD();
 - `grad(G,u,p)` or `G=grad(u,p)`: the gradient of `f` with respect to `u`
 - `hess(H,u,p)` or `H=hess(u,p)`: the Hessian of `f` with respect to `u`
 - `hv(Hv,u,v,p)` or `Hv=hv(u,v,p)`: the Hessian-vector product ``\frac{d^2 f}{du^2} v``.
-- `cons(res,x,p)` or `res=cons(x,p)`: the equality constraints vector, where the constraints
-  are satisfied when `res = 0`.
-- `cons_j(res,x,p)` or `res=cons_j(x,p)`: the Jacobian of the equality constraints.
-- `cons_h(res,x,p)` or `res=cons_h(x,p)`: the Hessian of the equality constratins, provided as
-  and array of Hessians with `res[i]` being the Hessian with respect to the `i`th output on `cons`.
+- `cons(x,p)` : the constraints function, should return a vector
+    with value of the `i`th constraint, evaluated at the current values of variables
+    inside the optimization routine. This takes just the function evaluations
+    and the equality or inequality assertion is applied by the solver based on the constraint
+    bounds passed as `lcons` and `ucons` to [OptimizationProblem](@ref), in case of equality
+    constraints `lcons` and `ucons` should be passed equal values.
+- `cons_j(res,x,p)` or `res=cons_j(x,p)`: the Jacobian of the constraints.
+- `cons_h(res,x,p)` or `res=cons_h(x,p)`: the Hessian of the constraints, provided as
+   an array of Hessians with `res[i]` being the Hessian with respect to the `i`th output on `cons`.
```

Comment:
This change is incorrect, which is what I think @gaurav-arya is confused about. We should have an in-place function for the constraints which corresponds to `OptimizationFunction{true}`, and then `res=cons(x,p)` should be the one created when `OptimizationFunction{false}`.

Basically, if `OptimizationFunction{true}`, then every single function should be mutating the first argument. `cons!(res,x,p)`, etc. The only exception is the `f` itself because it always returns a scalar (so there's no mutation possible). If `OptimizationFunction{false}`, then every function should return a vector.

Now I know that the solver backends have not been good at separating these interfaces (and generally use a combination of the two), but we should consider that an implementation detail for the solver wrappers. If a solver library only takes out of place versions for the constraints, then the wrapper should know that, check `if isinplace(prob)`, and if it is in place do `(x,p) -> f.cons(res,x,p); res)` with some `res` that it creates. If the solver only wants an in-place version, it should wrap the other direction. But the user shouldn't have to worry about this: these are implementation details about the solver that effect the performance of a solver, but every solver should try to work correctly (or error), with fully in-place definitions supporting a fully mutating non-allocating version (if any solvers support that), and fully out-of-place definitions supporting immutable arrays (if any solvers support that).